File: 45.jump-game-ii.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

public class Solution45
{
    // It's a BFS based solution. 
    // Each level is one jump.
    // Evertime, update the next round to get the maximum jump
    public int Jump(int[] nums) 
    {
        int res = 0, n = nums.Length, i = 0, cur = 0;
        while (cur < n - 1) 
        {
            ++res;
            int pre = cur;
            for (; i <= pre; ++i) {
                cur = Math.Max(cur, i + nums[i]);
            }
            if (pre == cur) return -1; // May not need this
        }
        return res;
    }
}

File: 45_jump-game-ii.py
from typing import List


class Solution:
    def jump(self, nums: List[int]) -> int:
        size = len(nums)
        reach = []
        steps = [size + 1] * size
        steps[0] = 0

        for i in range(size):
            for j in range(1, nums[i] + 1):
                if i+j < size:
                    steps[i + j]= min(steps[i + j], steps[i] + 1)
        
        return steps[size -1]

instance = Solution()
print(instance.jump([2,3,1,1,4]))

